File: services/cart-service/src/routes/health.js
const express = require('express');
const { sequelize } = require('../models');
const authService = require('../services/authService');
const productService = require('../services/productService');
const logger = require('../services/logger');

const router = express.Router();

// GET /health - Health check endpoint
router.get('/', async (req, res) => {
  try {
    // Test database connection
    await sequelize.authenticate();

    const healthResponse = {
      status: 'healthy',
      timestamp: new Date().toISOString(),
      database: 'connected',
    };

    res.json(healthResponse);
  } catch (error) {
    logger.error('Health check failed', error, {
      correlationId: req.correlationId,
    });

    res.status(503).json({
      status: 'unhealthy',
      timestamp: new Date().toISOString(),
      database: 'disconnected',
      error: 'Database connection failed',
    });
  }
});

// GET /health/ready - Readiness check endpoint
router.get('/ready', async (req, res) => {
  const checks = {
    database: false,
    auth_service: false,
    product_service: false,
  };

  try {
    // Check database
    await sequelize.authenticate();
    checks.database = true;
  } catch (error) {
    logger.warn('Database check failed', { correlationId: req.correlationId });
  }

  try {
    // Check auth service
    checks.auth_service = await authService.checkHealth();
  } catch (error) {
    logger.warn('Auth service check failed', {
      correlationId: req.correlationId,
    });
  }

  try {
    // Check product service
    checks.product_service = await productService.checkHealth();
  } catch (error) {
    logger.warn('Product service check failed', {
      correlationId: req.correlationId,
    });
  }

  const allHealthy = Object.values(checks).every((check) => check === true);
  const status = allHealthy ? 'ready' : 'not ready';
  const statusCode = allHealthy ? 200 : 503;

  res.status(statusCode).json({
    status,
    checks,
  });
});

module.exports = router;

File: services/user-profile-service/src/routers/health.py
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session
from sqlalchemy import text
from datetime import datetime

from ..database import get_db
from ..schemas.health import HealthResponse, ReadinessResponse, ReadinessChecks, HealthStatus, ReadinessStatus, DatabaseStatus
from ..services.auth_service import AuthService
from ..config import get_auth_service

router = APIRouter(tags=["Health"])


@router.get("/health", response_model=HealthResponse)
async def health_check(db: Session = Depends(get_db)):
    """Basic health check endpoint."""
    try:
        # Test database connection
        db.execute(text("SELECT 1"))
        database_status = DatabaseStatus.CONNECTED
        overall_status = HealthStatus.HEALTHY
    except Exception:
        database_status = DatabaseStatus.DISCONNECTED
        overall_status = HealthStatus.UNHEALTHY

    return HealthResponse(
        status=overall_status,
        service="user-profile-service",
        version="1.0.0",
        timestamp=datetime.utcnow(),
        database=database_status
    )


@router.get("/health/ready", response_model=ReadinessResponse)
async def readiness_check(
    db: Session = Depends(get_db),
    auth_service: AuthService = Depends(get_auth_service)
):
    """Readiness check with dependency verification."""
    checks = ReadinessChecks(
        database=ReadinessStatus.NOT_READY,
        auth_service=ReadinessStatus.NOT_READY
    )

    # Check database
    try:
        db.execute(text("SELECT 1"))
        checks.database = ReadinessStatus.READY
    except Exception:
        pass

    # Check auth service
    try:
        auth_healthy = await auth_service.check_auth_service_health()
        if auth_healthy:
            checks.auth_service = ReadinessStatus.READY
    except Exception:
        pass

    # Overall readiness
    overall_status = ReadinessStatus.READY if (
        checks.database == ReadinessStatus.READY and
        checks.auth_service == ReadinessStatus.READY
    ) else ReadinessStatus.NOT_READY

    return ReadinessResponse(
        status=overall_status,
        service="user-profile-service",
        checks=checks,
        timestamp=datetime.utcnow()
    )